torch/csrc/jit/export.cpp

```diff
@@ -883,6 +826,162 @@ void dump(const onnx::ModelProto& model, std::ostream& stream, size_t indent) {
+    }
+    for (auto& elem: module.get_methods()) {
+      torch::MethodDef* method_def = module_def->add_methods();
+      convertMethod(*elem.value(), method_def);
+    }
+    for (const auto& elem: module.get_modules()) {
+      torch::ModuleDef* sub_def = module_def->add_submodules();
+      convertModule(*elem->module, elem.key(), sub_def);
+    }
+  }
+  void convertParameter(const script::NamedParameter& param,
+      torch::ParameterDef* param_def) {
+    //std::cout << "param.name: " << param.name << "," << param.slot() << std::endl;
+    param_def->set_name(param.name);
+    param_def->set_is_buffer(param.is_buffer);
+    param_def->set_require_gradient(param.slot()->requires_grad());
+    //parameterMap_[param.slot()] = param.name;
+    convertTensor(*(param.slot()), param_def->mutable_tensor());
+  }
+  void convertTensor(const at::Tensor& tensor,
+      caffe2::TensorProto* tensor_proto) {
+    for (auto d : tensor.sizes()) {
+      tensor_proto->add_dims(d);
+    }
+    tensor_proto->set_data_type(
+        atenTypeToTensorProtoType(tensor.type().scalarType()));
+    tensor_proto->set_storage_type(caffe2::TensorProto_StorageType_EXTERNAL);
+    caffe2::ExternalDataProto* external_data = tensor_proto->mutable_external_data();
+    //std::cout << "strides: " << tensor.strides() << std::endl;
+    for (auto s : tensor.strides()) {
+      external_data->add_strides(s);
+    }
+    external_data->set_offset(tensor.storage_offset());
+    uint64_t record_size = tensor.type().elementSizeInBytes() * tensor.storage().size();
+    external_data->set_record_size(record_size);
+    auto* key = tensor.storage().unsafeGetStorageImpl();
+    auto it = storageMap_.find(key);
+    if (it == storageMap_.end()) {
+      // TODO HIP support
+      uint64_t record_id;
+      if (tensor.storage().device_type() == at::DeviceType::CUDA) {
+        // NB: This new tensor is created to support cuda tensors.
+        // Storages can be mutated when converting tensors from cuda to cpu,
+        // and we need a cpu tensor to copy data from.
+        at::Tensor t = at::getType(tensor)._th_tensor(
+            tensor.storage(),
+            /* storageOffset = */ 0,
+            /* size = */ { static_cast<int64_t>(tensor.storage().size()) },
+            /* stride = */ { 1 }).cpu();
+        AT_ASSERT(t.type().elementSizeInBytes() * t.storage().size() == record_size);
+        record_id = writer_.writeRecord(t.storage().data(),
+            t.type().elementSizeInBytes() * t.storage().size());
+        //std::cout << "===> record_id: " << record_id << std::endl;
+      } else {
+        record_id = writer_.writeRecord(tensor.storage().data(),
+            record_size);
+      }
+      external_data->set_record_id(caffe2::to_string(record_id));
+      storageMap_[key] = record_id;
+    } else {
+      external_data->set_record_id(caffe2::to_string(it->second));
+    }
+    // TODO handle device case, set the device_detail and load to CUDA device
```

Comment:
@gchanan - what is the preferred behavior of module.save / module.load in terms of CUDA devices. Should it preserve the device placement (if the module was CUDA()) or reset to CPU?

Reply:
We should mimic the python API. That means we should try to restore it to the same kind of device, and the same device number as it originally existed. The Python API also has some way to customize the loading process for when you are reloading to something that doesn't have a GPU. The default behavior should probably try to put the tensors back in the same place, but if the equivalent doesn't exist, to put it in the closest thing (e.g. GPU1 goes to GPU0 if there is only 1 GPU, and to CPU if there is no GPU at all).